### PR TITLE
WIP: Use Wurlitzer to capture stderr from the C extensions and display it …

### LINKED ISSUE #157: TypeError when Python C extension returns None
Whenever the Python C extension throws an error (e.g. due to input validation), `None` gets returned to the Python wrapper.  However, we then try to unpack that as a 2-tuple of `(extn_results, api_time)`, which results in a `TypeError` and the error message `'NoneType' object is not iterable`, which isn't helpful to the user.

This is less confusing on the command line when the error message from `stderr` is also immediately visible, but it's pretty confusing in a Jupyter notebook since the error message isn't captured and displayed there.

Here's an example that demonstrates the issue:
```python
import Corrfunc.theory.DD
import numpy as np

Corrfunc.theory.DD(1, 1, np.array([0., 1.]), np.array([1.]),np.array([1.]),np.array([1.]), boxsize=1.)
```
In a Notebook only the following is displayed:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-b20e8cca19cf> in <module>()
      2 import numpy as np
      3 
----> 4 Corrfunc.theory.DD(1, 1, np.array([0., 1.]), np.array([1.]),np.array([1.]),np.array([1.]), boxsize=1.)

~/Corrfunc/Corrfunc/theory/DD.py in DD(autocorr, nthreads, binfile, X1, Y1, Z1, weights1, periodic, X2, Y2, Z2, weights2, verbose, boxsize, output_ravg, xbin_refine_factor, ybin_refine_factor, zbin_refine_factor, max_cells_per_dim, c_api_timer, isa, weight_type)
    236                                      max_cells_per_dim=max_cells_per_dim,
    237                                      c_api_timer=c_api_timer,
--> 238                                      isa=integer_isa, **kwargs)
    239     if extn_results is None:
    240         msg = "RuntimeError occurred"

TypeError: 'NoneType' object is not iterable
```

The simple fix is to just check if the returned result is `None` before unpacking.  However, I propose we actually try to capture the `stderr` error message and display it to the user.  One way to do this would be with the [wurlitzer](https://github.com/minrk/wurlitzer) package.  [This section of DD.py](https://github.com/manodeep/Corrfunc/blob/master/Corrfunc/mocks/DDrppi_mocks.py#L363-L378) would then become:
```python
from wurlitzer import pipes

with pipes() as (out, err):
    extn_results = DD_extn(autocorr, nthreads, rbinfile,
                                     X1, Y1, Z1,
                                     periodic=periodic,
                                     verbose=verbose,
                                     boxsize=boxsize,
                                     output_ravg=output_ravg,
                                     xbin_refine_factor=xbin_refine_factor,
                                     ybin_refine_factor=ybin_refine_factor,
                                     zbin_refine_factor=zbin_refine_factor,
                                     max_cells_per_dim=max_cells_per_dim,
                                     c_api_timer=c_api_timer,
                                     isa=integer_isa, **kwargs)
if extn_results is None:
    msg = err.read()
    raise RuntimeError(msg)
else:
    extn_results, api_time = extn_results
```
Then you get the real error message, in a Notebook or command line:
```
RuntimeError: ../../utils/gridlink_impl_double.c> ERROR:  nlattice = 2 is so small that with periodic wrapping the same cells will be counted twice ....exiting
[...]
```

@manodeep, what do you think? wurlitzer is on pip for Python 2 and 3.

### PR DESCRIPTION
**Pull request by lgarrison**
…in Jupyter notebooks.  Resolves #157.

@manodeep, do you think you'll have time to test this?  It would be great to have independent verification that the command line behavior is unchanged and that error/progress messages show up in Jupyter.

Let's test on DD, then copy over to the other correlation statistics.

### CODE REVIEW

**Comment by lgarrison** [comment]
The astropy bot can't seem to find our changelog.  Do we need to rename it `CHANGES.rst`?  This suggests so: https://github.com/astropy/astropy-bot/blob/master/changebot/blueprints/changelog_helpers.py#L53

**Comment by manodeep** [comment]
I do want to test this out. But this might be a good opportunity to setup the bot and the any style guidelines we might want to enforce. That is, make the bot enforce for C/python style

I did start a `cformat` branch to take care of it but totally dropped the ball. (Also, that `cformat` branch was not supposed to be on github either)

**Comment by manodeep** [comment]
Everything else looks great! Let me test out the code on my laptop, and I will report back

File: Corrfunc/utils.py [hunk 1]
```diff
@@ -948,6 +948,30 @@ def is_native_endian(array):
...
+    We can use the Wurlitzer package to redirect stdout and stderr
+    from the command line into a Jupyter notebook.  But if we're not
+    in a notebook, this isn't safe because we can't redirect stdout
+    to itself.  This function is a thin wrapper that checks if the
+    stdout/err streams are TTYs and enables output redirection
+    based on that.
+
+    Basic usage is:
+
+    >>> with sys_pipes():
```
[the remark this answers is not in the document]
Comment (manodeep) [reply Corrfunc/utils.py]: Resolved by skype conversation that our code is so straight forward that doctests will not make sense

**Comment by manodeep** [comment]
I tested - works both in the command-line (with and without stdout/stderr redirects, redirects to the same file), and in a jupyter notebook

**Comment by lgarrison** [comment]
@manodeep I think that finishes everything I had planned for v2.2.  Should we make another release?  I think fixing the TypeError bug means it's probably worth it.

**Comment by manodeep** [comment]
@lgarrison The README is not rendering on `test.pypi.org`; and all the RST linter checks are implying everything is fine. Will continue to debug

**Comment by lgarrison** [comment]
I see that.  Weird!  We barely changed the readme.

**Comment by manodeep** [comment]
Resolved by removing the raw directive. Version 2.2.0 is now on PyPI

**Pull request merged**